File: dawn_cpp/ffbf.cpp
#include <bitset>
#include <functional>
#include <string>
#include <vector>

class BloomFilter {
private:
    std::bitset<1000000> bits;
    std::vector<std::function<std::size_t(std::string)>> hash_funcs;

public:
    BloomFilter(const std::vector<std::function<std::size_t(std::string)>>& hash_funcs)
        : hash_funcs(hash_funcs) {}

    void add(const std::string& element) {
        for (const auto& hash_func : hash_funcs) {
            bits.set(hash_func(element) % bits.size());
        }
    }

    bool contains(const std::string& element) const {
        for (const auto& hash_func : hash_funcs) {
            if (!bits.test(hash_func(element) % bits.size())) {
                return false;
            }
        }
        return true;
    }
};

File: dawn_go/ffbf.go
package main

import (
	"hash/fnv"
	"math/bits"
)

type BloomFilter struct {
	size      uint
	vector    []byte
	hashFuncs []uint64
}

func NewBloomFilter(size uint, hashFuncs []uint64) *BloomFilter {
	vector := make([]byte, (size+7)/8)
	return &BloomFilter{size, vector, hashFuncs}
}

func (bf *BloomFilter) Add(element string) {
	for _, hashFunc := range bf.hashFuncs {
		hash := fnv.New64a()
		hash.Write([]byte(element))
		bitIndex := uint(hash.Sum64()) % bf.size
		bf.vector[bitIndex/8] |= 1 << (bitIndex % 8)
	}
}

func (bf *BloomFilter) Contains(element string) bool {
	for _, hashFunc := range bf.hashFuncs {
		hash := fnv.New64a()
		hash.Write([]byte(element))
		bitIndex := uint(hash.Sum64()) % bf.size
		if (bf.vector[bitIndex/8] & (1 << (bitIndex % 8))) == 0 {
			return false
		}
	}
	return true
}